app/src/main/java/com/example/bottomnav/StructDecoder.java

```diff
@@ -0,0 +1,59 @@
+package com.example.bottomnav;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class StructDecoder implements DataDecoder {
+    private ArrayList<StructContents> contents;
+    public  HashMap<String, DataDecoder> decodedPrimatives = new HashMap<>();
+    private StringBuilder message = new StringBuilder();
+
+    public StructDecoder(ArrayList<StructContents> con) {
+        contents = con;
+    }
+
+    @Override
+    public String decode(Integer canId, byte[] payload) {
+        for (StructContents variable : contents) {
+            PayLoadDataType type = DataDecoder.lookup(variable.type);
+            DataDecoder primativeDecoder = DataDecoder.getPrimativeDecoder(type);
+            primativeDecoder.decode(canId, payload);
+            decodedPrimatives.put(variable.name, primativeDecoder);
+            payload = adjustPayload(payload, primativeDecoder.getPacketSize());
+        }
+        return getDecodedValues();
+    }
+
+    public byte[] adjustPayload(byte[] payload, int packetSize) {
+        int newSize = payload.length - packetSize;
+        ByteBuffer bb = ByteBuffer.wrap(payload);
+        if ((payload.length / packetSize) > 1) {
+            byte[] newPayload = new byte[newSize];
+            bb.get(newPayload, 0, packetSize);
+            bb.get(newPayload, 0, newSize);
+            return newPayload;
+        }
+        return payload;
+    }
+
+    public String getDecodedValues() {
+        for (int i = 0; i < contents.size(); i++) {
+            message.append("\n");
+            String varName = contents.get(i).name;
+            PayLoadDataType type = DataDecoder.lookup(contents.get(i).type);
+            message.append("    " + varName + ", " +
+                    DataDecoder.getPrimativeValue(type, decodedPrimatives.get(varName)));
+        }
+        return message.toString();
+    }
+
+    public DataDecoder getValue(String variableName) {
+        return decodedPrimatives.get(variableName);
+    }
+
+    @Override
+    public int getPacketSize() {
+        return 0;
```

Comment:
Make sure to actually implement these, or make sure they fail noisily (for example throwing a not implemented exception or something) - you don't want to leave submarine TODOs through your code.

The suggestion here would just be to sum the contents DataDecoder's getPacketSize,